File: src/2025-02-23-deepclaude-try/java/exercises/practice/book-store/src/main/java/BookStore.java
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;

public class BookStore {

    public double calculateBasketCost(List<Integer> books) {
        int[] counts = new int[5];
        for (int book : books) {
            counts[book - 1]++;
        }

        List<Integer> countsList = new ArrayList<>();
        for (int count : counts) {
            countsList.add(count);
        }

        double total = 0.0;
        List<Integer> groups = new ArrayList<>();

        while (true) {
            Collections.sort(countsList, Collections.reverseOrder());

            int groupSize = 0;
            while (groupSize < 5 && countsList.get(groupSize) > 0) {
                groupSize++;
            }

            if (groupSize == 0) {
                break;
            }

            groups.add(groupSize);

            double discount = 0.0;
            switch (groupSize) {
                case 5: discount = 0.25; break;
                case 4: discount = 0.20; break;
                case 3: discount = 0.10; break;
                case 2: discount = 0.05; break;
                case 1: discount = 0.00; break;
            }

            total += groupSize * 8 * (1 - discount);

            for (int i = 0; i < groupSize; i++) {
                countsList.set(i, countsList.get(i) - 1);
            }
        }

        int num5 = Collections.frequency(groups, 5);
        int num3 = Collections.frequency(groups, 3);
        int replacements = Math.min(num5, num3);
        total -= replacements * 0.40;

        return total;
    }

}

File: src/2025-02-23-deepclaude-try/go/exercises/practice/book-store/book_store.go
package bookstore

import (
	"sort"
)

func Cost(books []int) int {
	counts := make([]int, 5)
	for _, b := range books {
		counts[b-1]++
	}

	groups := make([]int, 6) // indexes 0-5 (0 unused)

	// Form groups greedily
	for {
		// Sort counts in descending order
		sort.Slice(counts, func(i, j int) bool {
			return counts[i] > counts[j]
		})

		// Find maximum possible group size
		maxGroup := 0
		for maxGroup < 5 && counts[maxGroup] > 0 {
			maxGroup++
		}

		if maxGroup == 0 {
			break
		}

		groups[maxGroup]++
		for i := 0; i < maxGroup; i++ {
			counts[i]--
		}
	}

	// Adjust groups: replace 5+3 with 4+4
	replacements := min(groups[5], groups[3])
	groups[5] -= replacements
	groups[3] -= replacements
	groups[4] += replacements * 2

	// Calculate total cost
	total := 0
	for k, cnt := range groups {
		switch k {
		case 1:
			total += cnt * 800
		case 2:
			total += cnt * (2 * 800 * 95 / 100)
		case 3:
			total += cnt * (3 * 800 * 90 / 100)
		case 4:
			total += cnt * (4 * 800 * 80 / 100)
		case 5:
			total += cnt * (5 * 800 * 75 / 100)
		}
	}

	return total
}

// Helper function to avoid using math.Min which uses float64
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
